az-cvm-vtpm/az-tdx-vtpm/src/hcl.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use crate::tdx::TdxVmReport;
+use jsonwebkey::JsonWebKey;
+use memoffset::offset_of;
+use serde::{Deserialize, Serialize};
+use serde_big_array::BigArray;
+use sev::firmware::guest::AttestationReport as SnpVmReport;
+use sha2::{Digest, Sha256};
+use std::mem;
+use thiserror::Error;
+
+const HCL_AKPUB_KEY_ID: &str = "HCLAkPub";
+const MAX_REPORT_SIZE: usize = mem::size_of::<SnpVmReport>();
+const MIN_REPORT_SIZE: usize = mem::size_of::<TdxVmReport>();
+
+#[derive(Error, Debug)]
+pub enum HclError {
+    #[error("invalid report type")]
+    InvalidReportType,
+    #[error("AkPub not found")]
+    AkPubNotFound,
+    #[error("parse error")]
+    ParseError(#[from] bincode::Error),
+}
+
+#[derive(Deserialize, Debug)]
+struct VarDataKeys {
+    keys: Vec<JsonWebKey>,
+}
+
+#[repr(u32)]
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+enum IgvmHashType {
+    Invalid = 0,
+    Sha256,
+    Sha384,
+    Sha512,
+}
+
+#[repr(u32)]
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
+enum IgvmReportType {
+    Invalid = 0,
+    VbsVm,
+    SnpVm,
+    Tvm,
+    TdxVm,
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+struct IgvmRequestData {
+    // Overall size of payload.
+    data_size: u32,
+    // Version of this structure, currently IGVM_ATTEST_VERSION_CURRENT (1).
+    version: u32,
+    report_type: IgvmReportType,
+    // The hash used in Report Data.
+    report_data_hash_type: IgvmHashType,
+    // Size of data blob.
+    variable_data_size: u32,
+    // Data holds AkPub or Transport Key.
+    variable_data: [u8; 0],
+}
+
+#[repr(u32)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+enum IgvmRequestType {
+    InvalidRequest = 0,
+    KeyReleaseRequest,
+    AkCertRequest,
+    KeyRotationInfoRequest,
+    KeyRotationStatus,
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+struct AttestationHeader {
+    signature: u32,
+    version: u32,
+    report_size: u32,
+    request_type: IgvmRequestType,
+    status: u32,
+    reserved: [u32; 3],
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+struct HwReport {
+    tdx_vm_report: TdxVmReport,
+    #[serde(with = "BigArray")]
+    _padding: [u8; MAX_REPORT_SIZE - MIN_REPORT_SIZE],
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+struct AttestationReport {
+    header: AttestationHeader,
+    hw_report: HwReport,
+    hcl_data: IgvmRequestData,
+}
+
+pub struct HclReport {
+    bytes: Vec<u8>,
+    attestation_report: AttestationReport,
+    report_type: ReportType,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum ReportType {
+    Tdx,
+    Snp,
+}
+
+impl HclReport {
+    pub fn new(bytes: Vec<u8>) -> Result<Self, HclError> {
```

Comment:
Any specific reason to go this route instead of `try_from` implementation for `HclReport`, as we do for `HclData` in `snp`'s `hcl`? I am not sure which is a better approach, but I was wondering? https://github.com/kinvolk/azure-cvm-tooling/blob/42d87db3ebaa9d37d25cb135bb849fba63427f94/az-snp-vtpm/src/hcl.rs#L123-L127

Reply:
it doesn't really matter, the fallible conversion is maybe a bit more idiomatic